tools/mason/MasonPublish.chpl

```diff
@@ -0,0 +1,263 @@
+line git commands if any of the git calls fails.*/
+proc publishPackage(username: string) throws {
+  const packageLocation = here.cwd();
+  var stream = makeRandomStream(int);
+  var uniqueDir = stream.getNext(): string;
+  const name = getPackageName();
+  const safeDir = (name + '-' + uniqueDir);
+  try! {
+    here.chdir(MASON_HOME);
+    if !exists('tmp') then mkdir('tmp');
+    here.chdir(MASON_HOME + '/tmp');
+    mkdir(safeDir);
+    here.chdir(safeDir);
+    cloneMasonReg(username);
+    here.chdir(packageLocation);
+    branchMasonReg(username, name, safeDir);
+    const package = addPackageToBricks(packageLocation, safeDir);
+    here.chdir(MASON_HOME + "/tmp/" + safeDir + "/mason-registry");
+    runCommand("git add .");
+    runCommand("git commit -m '" + package + "'");
+    runCommand('git push --set-upstream origin ' + package, true);
+    here.chdir(MASON_HOME + '/tmp');
+    rmTree(safeDir);
+    writeln('--------------------------------------------------------------------');
+    writeln('Go to the above link to open up a Pull Request to the mason-registry');
+  }
+  catch {
+    here.chdir(MASON_HOME + '/tmp');
+    rmTree(safeDir);
+    writeln('Error publishing your package to the mason-registry');
+  }
+}
+
+/* If --dry-run is passed then it takes the username and checks to see if the mason-registry is forked
+ and the package has a git remote origin. If both exist then the package can be published. */
+private proc dryRun(username: string) throws {
+  var fork = false;
+  var remoteCheck = checkIfForkExists(username: string);
+  if remoteCheck.exit_status == 0 {
+    fork = true;
+  }
+  var git = false;
+  if doesGitOriginExist() {
+    git = true;
+  }
+  if git && fork {
+    writeln('Package can be published to the mason-registry');
+    writeln('Commands that will be run:');
+    writeln('> git clone git:github.com:[username]/mason-registry mason-registry');
+    writeln('> git checkout -b [package name]');
+    writeln('Package Name will be added to the Bricks in the mason-registry');
+    writeln('> git add .');
+    writeln('> git commit -m [package name]');
+    writeln('> git push --set-upstream origin [package name]');
+  }
+  else {
+    if fork == false {
+      throw new owned MasonError('mason-registry is not forked on your GitHub');
+    }
+    else {
+      throw new owned MasonError('Package does not gave a git origin');
+    }
+  }
+}
+
+/* Opens Spawn call to get username for the mason registry fork */
+private proc usernameCheck(username: string) {
+  const gitRemote = 'git ls-remote https://github/com';
+  var usernameCheck = (gitRemote + username + '/mason-registry').split();
+  var p = spawn(usernameCheck);
+  p.wait();
+  return p;
+}
+/* Runs Commands to see if Fork of mason-registry exists under the username */
+private proc checkIfForkExists(username: string) {
+  var getFork= ('git ls-remote https://github.com/' + username + '/mason-registry').split();
+  var p = spawn(getFork, stdout=PIPE);
```

Comment:
is there a reason we are running spawn in some places and RunCommand in others?

Reply:
We need an exit status from this call which is easier to do with a direct spawn call. The others we do not need to capture an exit status

Reply:
Look for ``RunWithStatus`` within ``MasonUtils.chpl``. Should do what you want there.